c++/src/kj/io-test.c++

```diff
@@ -26,11 +26,75 @@
 #include "io.h"
 #include "debug.h"
 #include "miniposix.h"
+#include "test-util.h"
 #include <kj/compat/gtest.h>
 
 namespace kj {
 namespace {
 
+#if INTPTR_MAX == INT64_MAX
+static constexpr size_t ONE_GB = 1 * 1024 * 1024 * 1024;
+static constexpr size_t BUFFER_SIZE = 3 * ONE_GB;
+static constexpr kj::StringPtr ONE_GB_ALIGNED_PREFIXES_STORAGE[] = {
+// TODO(cleanup): KJ doesn't have an equivalent to std::array but if it were to that would be preferred.
+  "kj says hello 1"_kj,
+  "kj says hello 2"_kj,
+  "kj says hello 3"_kj,
+};
+static constexpr kj::ArrayPtr<const kj::StringPtr> ONE_GB_ALIGNED_PREFIXES{ONE_GB_ALIGNED_PREFIXES_STORAGE};
+static constexpr size_t PREFIX_SIZE_WITHOUT_NULL_TERMINATOR = ONE_GB_ALIGNED_PREFIXES[0].size();
+
+static const kj::StringPtr& nthPrefix(int gb) {
+  KJ_ASSERT(gb < ONE_GB_ALIGNED_PREFIXES.size());
+  return ONE_GB_ALIGNED_PREFIXES[gb];
+};
+
+static void setPrefix(char* ptr, int gb) {
+  KJ_ASSERT(gb < ONE_GB_ALIGNED_PREFIXES.size());
+  strcpy(ptr + gb * ONE_GB, (ONE_GB_ALIGNED_PREFIXES.begin() + gb)->cStr());
+};
+
+static void clearPrefix(char* ptr, int gb) {
+  memset(ptr + gb * ONE_GB, 0, (ONE_GB_ALIGNED_PREFIXES.begin() + gb)->size());
+};
+#endif
+
+TEST(Io, WriteLarge) {
```

Comment:
Is there an already existing preprocessor macro that protects running in release? If not I'll just #ifdef it out although that opens up this fix to potentially bitrot eventually.

Reply:
Unfortunately, no, there isn't an existing #ifdef to make things run only during the extended release tests. We'd need to add that in `super_test.sh`.

Reply:
This is what I have. Can't upload as the tests right now don't build on some platforms for some reason (i.e. don't want to accidentally submit broken code). Posted above the error message if you have thoughts on how I can resolve the issue.

```
diff --git a/c++/CMakeLists.txt b/c++/CMakeLists.txt
index 1a2ea097..0ef12e67 100644
--- a/c++/CMakeLists.txt
+++ b/c++/CMakeLists.txt
@@ -26,6 +26,7 @@ set(INSTALL_TARGETS_DEFAULT_ARGS
 # Options ======================================================================
 
 option(BUILD_TESTING "Build unit tests and enable CTest 'check' target." ON)
+option(BUILD_RELEASE_TESTING "Build & run a more expensive set of tests that should be run less frequently. Only takes effect when BUILD_TESTING is ON" OFF)
 option(EXTERNAL_CAPNP "Use the system capnp binary, or the one specified in $CAPNP, instead of using the compiled one." OFF)
 option(CAPNP_LITE "Compile Cap'n Proto in 'lite mode', in which all reflection APIs (schema.h, dynamic.h, etc.) are not included. Produces a smaller library at the cost of features. All programs built against the library must be compiled with -DCAPNP_LITE. Requires EXTERNAL_CAPNP." OFF)
 
diff --git a/c++/src/kj/CMakeLists.txt b/c++/src/kj/CMakeLists.txt
index 8f0dbb3b..797c48f0 100644
--- a/c++/src/kj/CMakeLists.txt
+++ b/c++/src/kj/CMakeLists.txt
@@ -238,6 +238,9 @@ if(BUILD_TESTING)
   target_link_libraries(kj-tests kj-test kj)
   add_dependencies(check kj-tests)
   add_test(NAME kj-tests-run COMMAND kj-tests)
+  if(BUILD_RELEASE_TESTING)
+    target_compile_definitions(kj-tests PRIVATE -DKJ_TEST_LARGE_IO=1)
+  endif()
 
   if(NOT CAPNP_LITE)
     add_executable(kj-heavy-tests
diff --git a/c++/src/kj/io-test.c++ b/c++/src/kj/io-test.c++
index 7a16b3aa..1677cf2f 100644
--- a/c++/src/kj/io-test.c++
+++ b/c++/src/kj/io-test.c++
@@ -77,6 +77,7 @@ static void setPrefix(char* ptr, int gb) {
 };
 #endif
 
+#ifdef KJ_TEST_LARGE_IO
 TEST(Io, WriteLarge) {
   // Check that writing a single large write > 2GB succeeds. This is primarily to workaround an
   // Apple POSIX compliance issue (filed to Apple as FB8934446). This test only makes sense on
@@ -111,6 +112,7 @@ TEST(Io, WriteLarge) {
   }
 #endif
 }
+#endif  // KJ_TEST_LARGE_IO
 
 TEST(Io, WriteVec) {
   // Check that writing an array of arrays works even when some of the arrays are empty.  (This
@@ -139,6 +141,7 @@ TEST(Io, WriteVec) {
   EXPECT_STREQ("foobar", buf);
 }
 
+#ifdef KJ_TEST_LARGE_IO
 TEST(Io, WriteVecLarge) {
   // Check that writing an array of arrays works when one a single writev entry is > 2GB. This is
   // primarily to workaround an Apple POSIX compliance issue (filed to Apple as FB8934446). This
@@ -187,6 +190,7 @@ TEST(Io, WriteVecLarge) {
   }
 #endif
 }
+#endif // KJ_TEST_LARGE_IO
 
 KJ_TEST("stringify AutoCloseFd") {
   int fds[2];
diff --git a/super-test.sh b/super-test.sh
index 08d072e0..fe9bc02b 100755
--- a/super-test.sh
+++ b/super-test.sh
@@ -180,7 +180,7 @@ while [ $# -gt 0 ]; do
       rm -rf cmake-build
       mkdir cmake-build
       cd cmake-build
-      doit cmake -G "Unix Makefiles" ..
+      doit cmake -G "Unix Makefiles" -DBUILD_RELEASE_TESTING=ON ..
       doit make -j$PARALLEL check
       exit 0
       ;;

```

My thought is that rather than totally compiling out the tests as I do now, I should just somehow make them "no-ops" by doing an early return so that at least the code can't bitrot syntactically (unless the release stuff is tested in some automated system regularly that would flag it on a breakage?).

Reply:
Release tests run in CI but only after merging to master, so I wouldn't worry too much about bitrotting. Though I wonder if the CI machines have enough RAM and disk space available to handle this...

I'd use the autotools build rather than cmake. Otherwise if this test fails it'll look like a cmake-specific problem and people will be confused. autotools is our primary build system. Plus with autotools, you can directly pass CXXFLAGS to `configure` so you don't need to update the build files at all.